htlcswitch/link_test.go

```diff
@@ -1110,7 +1110,6 @@ func TestChannelLinkMultiHopUnknownPaymentHash(t *testing.T) {
 	// Send payment and expose err channel.
 	err = n.aliceServer.htlcSwitch.SendHTLC(
 		n.firstBobChannelLink.ShortChanID(), pid, htlc,
-		newMockDeobfuscator(),
```

Comment:
Is this movement really needed in order to complete the reliable payments feature? The onion decryption/encryption portion is a _link level_ detail, meaning that the router doesn't need to be aware of it at all. The current placement is by design, in that the router crafts the route, but all error encryption/decryption is handled by the switch. As a result, the router gets back a nice concrete error, and doesn't need to know about the details of error encryption/decryption. 

IMO, it's a code smell that the router is now _creating_ then _returning_ an `htlcswitch.ForwardingError` back to itself. It's also the case that we already store the items required to reconstruct the circuit (for encrypting and decrypting errors) in the switch's circuit map. Additionally, there's some particular context in the switch w.r.t things like if this is a locally sourced error, a resolution, and so on that are required in order to properly handle all classes of errors. Overall, I don't see why this change is warranted.

Reply:
As pointed out in https://github.com/lightningnetwork/lnd/pull/2762/#discussion_r281499203, the current placement is awkward, as _both_ the `router` (encryption) and `switch` (decryption) is aware of the onion format now. By moving it all into either one, it should be less shared responsibility.

I moved it all into the `router` since it has everything that is needed to decrypt the error (route + session key) by recreating the `ErrorDecryptor` after a restart. We could possibly move it all into the `switch`, but that would require us to store this information across restarts, as it is currently not (the curcuit map only stores onion blobs from what I can tell). If that's what we want then a natural place to put it would be the circuit map, as part of the `paymentCircuit`?

And finally, the reason this is needed for reliable payments is to handle received errors across restarts. The `switch` will already handle persisting received results (see #2762), but it cannot deobfuscate the errors. This has to be done by the router, either by getting the obfuscated data from the switch, or by giving it the `deobfuscator` to the switch after a restart (but this is a bit tricky timingwise).

Reply:
> or by giving it the deobfuscator to the switch after a restart (but this is a bit tricky timingwise).
I made a rough attempt doing this, and the diff is not that bad:  https://github.com/lightningnetwork/lnd/commit/9e2c428f34c6de9486fceb4d09a376671b4e578c

lmk what is preferred.

Reply:
Went with the smallest diff case.